pkg/logical-cluster-manager/cluster-manager-controller.go

```diff
@@ -0,0 +1,244 @@
+		kubeconfig:       kubeconfig,
+		ctx:              ctx,
+		clusterclientset: clusterclientset,
+		clusterInformer:  clusterInformer,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			"cluster-controller"),
+	}
+
+	logger.Info("Setting up event handlers")
+	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			logger.Info("Add cluster")
+			controller.queue.Add(
+				queueItem{
+					op:  opAdd,
+					key: obj,
+				},
+			)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Most fields in the logical cluster are immutable with the
+			// exception of the URL and fields in the status.
+			oldInfo := oldObj.(*v1alpha1apis.LogicalCluster)
+			newInfo := newObj.(*v1alpha1apis.LogicalCluster)
+			if !reflect.DeepEqual(oldInfo.Status, newInfo.Status) {
+				controller.queue.Add(
+					queueItem{
+						op:  opUpdate,
+						key: newObj,
+					},
+				)
+			}
+		},
+		DeleteFunc: func(delObj interface{}) {
+			logger.Info("Delete cluster")
+			controller.queue.Add(
+				queueItem{
+					op:  opDelete,
+					key: delObj,
+				},
+			)
+		},
+	})
+
+	logger.Info("New %s controller", controllerName)
+	return controller
+}
+
+// Run starts the controller, which stops when c.context.Done() is closed.
+func (c *Controller) Run(numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := klog.FromContext(c.ctx)
+	logger.Info("starting controller")
+	defer logger.Info("shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(c.ctx, c.runWorker, time.Second)
+	}
+
+	<-c.ctx.Done()
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	i, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(i)
+
+	//ES: not sure we should treat all errors in the same way (queue back)
+	if err := c.process(ctx, i.(queueItem)); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller didn't sync, err: %w", controllerName, err))
+		c.queue.AddRateLimited(i)
+		return true
+	}
+	c.queue.Forget(i)
+	return true
+}
+
+func (c *Controller) processAdd(ctx context.Context, key any) error {
+	logger := klog.FromContext(ctx)
+	var err error
+
+	newClusterConfig := key.(*v1alpha1apis.LogicalCluster)
+	clusterName := newClusterConfig.Spec.ClusterName
+
+	provider := clusterproviderclient.GetProviderClient(newClusterConfig.Spec.ProviderType, newClusterConfig.Spec.ProviderName)
+
+	// Update status to NotReady
+	newClusterConfig.Status.Phase = v1alpha1apis.LogicalClusterPhaseNotReady
+	_, err = c.clusterclientset.EdgeV1alpha1().LogicalClusters().Update(ctx, newClusterConfig, v1.UpdateOptions{})
```

Comment:
What does a client that queries the status before this line runs get?  How/why is that semantically different from LogicalClusterPhaseNotReady?